driver-async/src/main/com/mongodb/async/client/gridfs/GridFSDownloadStreamImpl.java

```diff
@@ -122,61 +122,87 @@ public void onResult(final GridFSFile result, final Throwable t) {
+                            }
+                        }
+                    });
+        } else {
+            cursor.next(new SingleResultCallback<List<Document>>() {
                 @Override
-                public void onResult(final byte[] result, final Throwable t) {
+                public void onResult(final List<Document> result, final Throwable t) {
                     if (t != null) {
-                        releaseReadingLock();
                         callback.onResult(null, t);
+                    } else if (result == null || result.isEmpty()) {
+                        chunkNotFound(chunkIndex, callback);
                     } else {
-                        buffer = result;
-                        readFromBuffer(dst, recursiveCallback);
+                        resultsQueue.addAll(result);
+                        if (batchSize == 1) {
+                            discardCursor();
+                        }
+                        processResults(amountRead, dst, callback);
                     }
                 }
             });
-        } else {
-            readFromBuffer(dst, recursiveCallback);
         }
     }
 
-    private void readFromBuffer(final ByteBuffer dst, final SingleResultCallback<Void> callback) {
-        int amountToCopy = dst.remaining();
-        if (amountToCopy > buffer.length - bufferOffset) {
-            amountToCopy = buffer.length - bufferOffset;
+    private void processResults(final int previousAmountRead, final ByteBuffer dst, final SingleResultCallback<Integer> callback) {
+        ChunkDataOrError chunkOrError = new ChunkDataOrError();
+        int amountRead = previousAmountRead;
+        while (currentPosition < fileInfo.getLength() && dst.remaining() > 0 && !resultsQueue.isEmpty()) {
+            if (buffer == null || bufferOffset == buffer.length) {
+                getBufferFromChunk(resultsQueue.poll(), chunkIndex, chunkOrError);
+                if (chunkOrError.hasError()) {
+                    callback.onResult(null, chunkOrError.getError());
+                    return;
+                } else {
+                    buffer = chunkOrError.getBuffer();
+                    bufferOffset = 0;
+                    chunkIndex += 1;
+                }
+            }
+
+            int amountToCopy = dst.remaining();
+            if (amountToCopy > buffer.length - bufferOffset) {
+                amountToCopy = buffer.length - bufferOffset;
+            }
+            dst.put(buffer, bufferOffset, amountToCopy);
+            bufferOffset += amountToCopy;
+            currentPosition += amountToCopy;
+            amountRead += amountToCopy;
         }
-        dst.put(buffer, bufferOffset, amountToCopy);
-        bufferOffset += amountToCopy;
-        currentPosition += amountToCopy;
-        callback.onResult(null, null);
+
+        checkAndFetchResults(amountRead, dst, callback);
     }
 
     @Override
```

Comment:
I know it's not part of the PR, but when running with the code coverage tool, I noticed that the close method is never called. Can we add a test for it?